File: apps/ios/Sources/HelpCore/DayWords.swift
// How a day is said to a person: "Today", "Tomorrow", or "Friday, Sep 25" — never "2026-09-25". The same wording
// as `dayName` in apps/web/src/main.ts, so a closed listing reads the same on the web and on a phone.
//
// Pure: the caller hands in the day it is in Detroit, the two words from strings/*.json (`day.today`,
// `day.tomorrow`) and the locale, so `swift test` holds it on Linux with no clock and no bundle.
import DetroitQuery
import Foundation

public enum DayWords {
    public enum Kind: Equatable, Sendable { case today, tomorrow, other }

    /// `date` and `today` are calendar days, "YYYY-MM-DD" (a longer ISO string is read by its first ten characters).
    public static func kind(date: String, today: String) -> Kind? {
        guard let d = parseDay(date), let t = parseDay(today) else { return nil }
        return d == t ? .today : d == t + 1 ? .tomorrow : .other
    }

    /// The Detroit calendar day of an instant, "YYYY-MM-DD".
    public static func detroitDay(of instant: Date) -> String {
        let w = toWall(instant)
        return dayString(dayNumber(w.y, w.m, w.d))
    }

    /// A day that cannot be read is handed back as it came: a wrong date is worse than an odd-looking one.
    public static func name(date: String, today: String, locale: Locale,
                            todayWord: String, tomorrowWord: String) -> String {
        guard let kind = kind(date: date, today: today), let day = parseDay(date) else { return date }
        switch kind {
        case .today: return todayWord
        case .tomorrow: return tomorrowWord
        case .other:
            // A calendar day has no time zone: it is formatted in UTC from a UTC midnight, as the web does.
            let f = DateFormatter()
            f.locale = locale
            f.timeZone = TimeZone(identifier: "UTC")
            f.setLocalizedDateFormatFromTemplate("EEEEMMMd")
            return f.string(from: Date(timeIntervalSince1970: Double(day) * 86400))
        }
    }
}

File: apps/android/app/src/main/kotlin/org/help313/app/DayWords.kt
// How a day is said to a person: "Today", "Tomorrow", or "Friday, Sep 25" — never "2026-09-25", and never a year.
// The same wording as `dayName` in apps/web/src/main.ts and as apps/ios/Sources/HelpCore/DayWords.swift, so a closed
// listing reads the same on the web and on both phones.
//
// Pure, and no android.* class: the caller hands in the day it is in Detroit, the two words from the strings files
// (`day.today`, `day.tomorrow`) and the locale, so `:core:test` holds it on a plain JDK with no clock and no bundle.
//
// The weekday is the locale's own LONG name. The strings files carry only the short forms (`day.MO` is "Mon"), which
// are for the hours table; the web and the iPhone both print the long one here.
package org.help313.app

import org.help313.query.parseDay
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

object DayWords {

    enum class Kind { TODAY, TOMORROW, OTHER }

    /** [date] and [today] are calendar days, "YYYY-MM-DD". Null when either cannot be read. */
    fun kind(date: String, today: String): Kind? {
        val d = parseDay(date) ?: return null
        val t = parseDay(today) ?: return null
        return if (d == t) Kind.TODAY else if (d == t + 1) Kind.TOMORROW else Kind.OTHER
    }

    /**
     * Weekday, month, day — in the order that language writes them. `java.text` has no "best pattern for this
     * skeleton" on a plain JDK (that is `android.text.format.DateFormat`, which `:core` cannot see), so the four
     * orders are written down; the names themselves are the locale's own.
     */
    private fun pattern(locale: Locale): String = when (locale.language) {
        "en" -> "EEEE, MMM d"
        "ar" -> "EEEE، d MMM"
        else -> "EEEE, d MMM"
    }

    /** "Friday, Sep 25". A calendar day has no time zone: it is formatted in UTC from a UTC midnight, as the web does. */
    fun long(date: String, locale: Locale): String {
        val day = parseDay(date) ?: return date
        val f = SimpleDateFormat(pattern(locale), locale)
        f.timeZone = TimeZone.getTimeZone("UTC")
        return f.format(Date(day.toLong() * 86_400_000L))
    }

    /** A day that cannot be read is handed back as it came: a wrong date is worse than an odd-looking one. */
    fun name(date: String, today: String, locale: Locale, todayWord: String, tomorrowWord: String): String =
        when (kind(date, today)) {
            Kind.TODAY -> todayWord
            Kind.TOMORROW -> tomorrowWord
            Kind.OTHER -> long(date, locale)
            // No "today" to compare with: still a day by name, never the raw ISO date.
            null -> long(date, locale)
        }
}
